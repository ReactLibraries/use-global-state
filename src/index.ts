import {
  createContext,
  createElement,
  Dispatch,
  SetStateAction,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

type Render<T = unknown> = Dispatch<SetStateAction<T | undefined>>;
export type ContextType = {
  renderMap: Map<string, Set<Render>>;
  initialKeys: Set<string>;
  cache: { [key: string]: unknown };
};

const context = createContext<ContextType>(undefined as never);
export const Provider = ({
  value,
  children,
}: {
  value?: Partial<ContextType>;
  children?: React.ReactNode;
}) => {
  const v = value || {};
  v.renderMap = new Map<string, Set<Render>>();
  v.initialKeys = new Set<string>();
  v.cache = {};
  return createElement<{ value: ContextType }>(
    context.Provider,
    { value: v as ContextType },
    children
  );
};
const NormalizeKey = (keys: string | string[]) =>
  (Array.isArray(keys) ? keys : [keys]).reduce((a, b) => `${a}[${b}]`, '');

const globalContext = {
  renderMap: new Map<string, Set<Render>>(),
  initialKeys: new Set<string>(),
  cache: {},
};

export const reset = (context: ContextType = globalContext) => {
  const { renderMap, initialKeys, cache } = context;
  renderMap.clear();
  initialKeys.clear();
  Object.keys(cache).forEach((key) => delete cache[key]);
};

export const getCache = <T = unknown>(
  keys: string | string[],
  context: ContextType = globalContext
) => {
  const { cache } = context;
  const key = NormalizeKey(keys);
  const result: { [key: string]: T } = {};
  Object.entries(cache)
    .filter(([k]) => k.indexOf(key) === 0)
    .forEach(([key, value]) => (result[key] = value as T));
  return result;
};
export const setCache = <T = unknown>(
  src: { [key: string]: T },
  context: ContextType = globalContext
) => {
  const { initialKeys, cache } = context;
  Object.entries(src).forEach(([key, value]) => {
    cache[key] = value as T;
    initialKeys.add(key);
  });
};

export const query = <T = unknown>(keys: string | string[], context: ContextType = globalContext) =>
  context.cache[NormalizeKey(keys)] as T;

export const mutate = <T = Object>(
  keys: string | string[],
  data: T | Promise<T> | ((data: T) => T | Promise<T>),
  context: ContextType = globalContext
) => {
  const { renderMap, initialKeys, cache } = context;
  const key = NormalizeKey(keys);
  const value =
    typeof data === 'function' ? (data as (data: T) => T | Promise<T>)(cache[key] as T) : data;
  if (value instanceof Promise) {
    value.then((data) => {
      cache[key] = data;
      renderMap.get(key)?.forEach((render) => render(data));
      initialKeys.add(key);
    });
  } else {
    cache[key] = data;
    renderMap.get(key)?.forEach((render) => render(data));
    initialKeys.add(key);
  }
};

export const useGlobalState: {
  <T>(keys: string | string[], initialData: T | (() => T)): readonly [
    T,
    (data: T | ((data: T) => T)) => void
  ];
  <K = unknown, T = K | undefined>(keys: string | string[]): readonly [
    T,
    (data: T | ((data: T) => T)) => void
  ];
} = <T>(keys: string | string[], initialData?: T | (() => T)) => {
  const c = useContext<ContextType>(context) || globalContext;
  const { renderMap, initialKeys, cache } = c;

  type RenderMap = Map<string, Set<Render<T>>>;
  const key = useMemo(() => NormalizeKey(keys), Array.isArray(keys) ? keys : [keys]);
  const [state, render] = useState<T | undefined>((cache[key] as T) || initialData);
  const dispatch = useCallback(
    (data: T | ((data: T) => T)) => {
      mutate(keys, data, c);
    },
    [key]
  );
  let init = false;
  useEffect(() => {
    const renders = renderMap.get(key)!;
    init && renders.forEach((r) => r(cache[key]));
    return () => {
      (renderMap as RenderMap).get(key)?.delete(render);
    };
  }, []);
  if (initialData !== undefined && !initialKeys.has(key)) {
    cache[key] = typeof initialData === 'function' ? (initialData as () => T)() : initialData;
    initialKeys.add(key);
    init = true;
  }
  const renders = (renderMap as RenderMap).get(key)!;
  if (renders) renders.add(render);
  else (renderMap as RenderMap).set(key, new Set<Render<T>>().add(render));

  return [state, dispatch] as const;
};
