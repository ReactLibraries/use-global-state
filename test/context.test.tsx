import React, { useEffect, useState } from 'react';
import { createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import { createContextCache, reset, useGlobalState, useMutation } from '../src/index';
import { Provider } from '../src/index';

let container: HTMLElement;
beforeEach(() => {
  reset();
  container = document.createElement('div');
  document.body.appendChild(container);
  global.IS_REACT_ACT_ENVIRONMENT = true;
});

afterEach(() => {
  container.remove();
});

it('initData-undefined', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider>
        <Component01 />
        <Component02 />
        <Component03 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('initData-normal', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    //first initData
    const [value] = useGlobalState('Key1', 1);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1', 2);
    return <>{value ?? 'undefined'}</>;
  };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider>
        <Component01 />
        <Component02 />
        <Component03 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('initData-function', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    //first initData
    const [value] = useGlobalState('Key1', () => 1);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1', () => 2);
    return <>{value ?? 'undefined'}</>;
  };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider>
        <Component01 />
        <Component02 />
        <Component03 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('useEffect', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    const [value, setValue] = useGlobalState('Key1', 1);
    useEffect(() => {
      setValue(3);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value, setValue] = useGlobalState<unknown>('Key1', 2);
    useEffect(() => {
      //Last setValue
      setValue('Effect');
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider>
        <Component01 />
        <Component02 />
        <Component03 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('onclick', () => {
  const Component01 = () => {
    const [value, setValue] = useGlobalState('Key1');
    return (
      <button
        id="click"
        onClick={() => {
          setValue('Click');
        }}
      >
        <>{value ?? 'undefined'}</>
      </button>
    );
  };
  const Component02 = () => {
    const [value, setValue] = useGlobalState<unknown>('Key1', 1);
    useEffect(() => {
      setValue('Efect');
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1', 2);
    return <>{value ?? 'undefined'}</>;
  };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider>
        <Component01 />
        <Component02 />
        <Component03 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    const element = container.querySelector('#click');
    element?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('remove', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    //first initData
    const [value] = useGlobalState('Key1', () => 1);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1', () => 2);
    return <>{value ?? 'undefined'}</>;
  };
  const Root = () => {
    const [flag, setFlag] = useState(true);
    useEffect(() => {
      setFlag(false);
    }, []);

    return (
      <Provider>
        <Component01 />
        <Component02 />
        {flag && <Component03 />}
      </Provider>
    );
  };
  const root = createRoot(container);
  act(() => {
    root.render(<Root />);
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('Provider value', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1', 5);
    return <>{value}</>;
  };
  const value = [['Key1', 10]] as const;
  let newValue: { [key: string]: unknown }, oldVallue: { [key: string]: unknown };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider value={createContextCache(value)}>
        <Component01 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  expect([oldVallue, newValue]).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('Provider update', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1', 5);
    const mutate = useMutation();
    useEffect(() => {
      mutate('Key2', 'ABC');
    }, []);
    return <>{value}</>;
  };
  let newValue: { [key: string]: unknown }, oldVallue: { [key: string]: unknown };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider
        onUpdate={(nv, ov) => {
          newValue = nv;
          oldVallue = ov;
        }}
      >
        <Component01 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  expect([oldVallue, newValue]).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('Provider nest update', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1', 5);
    const mutate = useMutation();
    useEffect(() => {
      mutate('Key2', 'ABC');
    }, []);
    return <>{value}</>;
  };
  let newValue: { [key: string]: unknown }, oldVallue: { [key: string]: unknown };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider
        onUpdate={(nv, ov) => {
          newValue = nv;
          oldVallue = ov;
        }}
      >
        <Provider>
          <Component01 />
        </Provider>
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  expect([oldVallue, newValue]).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});

it('Provider unmount', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1', 5);
    const [state, setState] = useState(true);
    useEffect(() => {
      setState(false);
    }, []);
    return state && <Provider>{value}</Provider>;
  };
  let newValue: { [key: string]: unknown }, oldVallue: { [key: string]: unknown };
  const root = createRoot(container);
  act(() => {
    root.render(
      <Provider
        onUpdate={(nv, ov) => {
          newValue = nv;
          oldVallue = ov;
        }}
      >
        <Component01 />
      </Provider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  expect([oldVallue, newValue]).toMatchSnapshot();
  act(() => {
    root.unmount();
  });
});
