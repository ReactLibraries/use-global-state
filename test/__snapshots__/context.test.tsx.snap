// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider value 1`] = `
NodeList [
  10,
]
`;

exports[`initData-function 1`] = `
NodeList [
  1,
  1,
  1,
]
`;

exports[`initData-normal 1`] = `
NodeList [
  1,
  1,
  1,
]
`;

exports[`initData-undefined 1`] = `
NodeList [
  undefined,
  undefined,
  undefined,
]
`;

exports[`onclick 1`] = `
NodeList [
  <button
    id="click"
  >
    Efect
  </button>,
  Efect,
  Efect,
]
`;

exports[`onclick 2`] = `
NodeList [
  <button
    id="click"
  >
    Click
  </button>,
  Click,
  Click,
]
`;

exports[`remove 1`] = `
NodeList [
  1,
  1,
]
`;

exports[`useEffect 1`] = `
NodeList [
  Effect,
  Effect,
  Effect,
]
`;
